ast/src/analyzed/types.rs

```diff
@@ -1,287 +0,0 @@
-            Type::Array(_) | Type::Tuple(_) => false,
-            Type::Function(_) => true,
-            Type::TypeVar(_) => false,
-            _ => unreachable!(),
-        }
-    }
-
-    pub fn is_concrete_type(&self) -> bool {
-        self.contained_type_vars_with_repetitions().next().is_none()
-    }
-
-    pub fn contains_type_var(&self, name: &str) -> bool {
-        self.contained_type_vars_with_repetitions()
-            .any(|n| n == name)
-    }
-
-    /// Returns the list of contained type vars in order of first occurrence.
-    pub fn contained_type_vars(&self) -> impl Iterator<Item = &String> {
-        self.contained_type_vars_with_repetitions().unique()
-    }
-
-    /// Substitutes all occurrences of the given type variables with the given types.
-    /// Does not apply the substitutions inside the replacements.
-    pub fn substitute_type_vars(&mut self, substitutions: &HashMap<String, Type>) {
-        match self {
-            Type::TypeVar(n) => {
-                if let Some(t) = substitutions.get(n) {
-                    *self = t.clone();
-                }
-            }
-            Type::Array(ArrayType { base, length: _ }) => {
-                base.substitute_type_vars(substitutions);
-            }
-            Type::Tuple(TupleType { items }) => {
-                items
-                    .iter_mut()
-                    .for_each(|t| t.substitute_type_vars(substitutions));
-            }
-            Type::Function(FunctionType { params, value }) => {
-                params
-                    .iter_mut()
-                    .for_each(|t| t.substitute_type_vars(substitutions));
-                value.substitute_type_vars(substitutions);
-            }
-            _ => {
-                assert!(self.is_elementary());
-            }
-        }
-    }
-
-    pub fn substitute_type_vars_to(mut self, substitutions: &HashMap<String, Type>) -> Self {
-        self.substitute_type_vars(substitutions);
-        self
-    }
-
-    fn contained_type_vars_with_repetitions(&self) -> Box<dyn Iterator<Item = &String> + '_> {
-        match self {
-            Type::TypeVar(n) => Box::new(std::iter::once(n)),
-            Type::Array(ar) => ar.base.contained_type_vars_with_repetitions(),
-            Type::Tuple(tu) => Box::new(
-                tu.items
-                    .iter()
-                    .flat_map(|t| t.contained_type_vars_with_repetitions()),
-            ),
-            Type::Function(fun) => Box::new(
-                fun.params
-                    .iter()
-                    .flat_map(|t| t.contained_type_vars_with_repetitions())
-                    .chain(fun.value.contained_type_vars_with_repetitions()),
-            ),
-            _ => {
-                assert!(self.is_elementary());
-                Box::new(std::iter::empty())
-            }
-        }
-    }
-}
-
-impl<Ref: Display> From<TypeName<Expression<Ref>>> for Type {
-    fn from(value: TypeName<Expression<Ref>>) -> Self {
```

Comment:
This conversion is retained by having `impl From<Type<Expression<Ref>>> for Type<u64>`.